dali_tf_plugin/dali_dataset_op.cc

```diff
@@ -274,10 +251,31 @@ class DALIDatasetOp : public DatasetOpKernel {
             enable_memory_stats_(enable_memory_stats) {}
 
       Status Initialize(IteratorContext* context) override {
-        if (!dataset()->fail_on_device_mismatch_) {
-            LOG(WARNING) << "DALI LOG: Allocator Name in Iterator: " << context->allocator({})->Name();
+        return CheckOutputDevices(); 
+      }
+
+      Status CheckOutputDevices() {
+        auto num_outputs = daliGetNumOutput(&pipeline_handle_);
+        for (auto i  = 0; i < num_outputs; ++i) {
+          auto dali_device_type = daliGetOutputDevice(&pipeline_handle_, i);
+
+          if (dali_device_type != dataset()->device_type_) {
+            std::stringstream msg;
+            msg << "TF device and DALI device mismatch. TF device: ";
+            msg << (dataset()->device_type_ == device_type_t::CPU ? "CPU" : "GPU");
+            msg << ", DALI device: ";
+            msg << (dali_device_type == device_type_t::CPU ? "CPU" : "GPU");
+            msg << " for output " << i;
```

Comment:
Use make_string? Not sure how this would affect readability but that's more of a DALI way of doing this.

Reply:
Done